File: ai/labs/lab2/fuzzy set/lab2_fuzzy_set/RatingFuzzySet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab2_fuzzy_set
{
    public class RatingFuzzySet
    {
        protected Dictionary<eRating, eAdvantages> fzRating;
        protected ArrayList vBelongValues;
        protected List<List<float>> liFzMatrix;
        
        public List<eRating> createRatingList()
        {
            List<eRating> liRatings = new List<eRating>();
            liRatings.Add(eRating.LOW);
            liRatings.Add(eRating.LIGHT_MIDDLE);
            liRatings.Add(eRating.MIDDLE);
            liRatings.Add(eRating.LIGHT_HIGH);
            liRatings.Add(eRating.HIGH);
            liRatings.Add(eRating.MAXIMAL);
            return liRatings;
        }
        public void setRatingSet()
        {
            vBelongValues = new ArrayList();
            fzRating = new Dictionary<eRating, eAdvantages>();
            fzRating.Add(eRating.LOW, eAdvantages.ABSOLUTE);
            fzRating.Add(eRating.LIGHT_MIDDLE, eAdvantages.INTERMEDIATE_ABSOLUTE);
            fzRating.Add(eRating.MIDDLE, eAdvantages.APPARENT);
            fzRating.Add(eRating.LIGHT_HIGH, eAdvantages.ESSENTIAL);
            fzRating.Add(eRating.HIGH, eAdvantages.WEAK);
            fzRating.Add(eRating.MAXIMAL, eAdvantages.APSEND);
        }

        public List<List<float>> buildMatrix()
        {
            liFzMatrix = new List<List<float>>();
            List<eRating> liRatings = this.createRatingList();
            liFzMatrix.Add(new List<float>(fzRating.Count));
            for (int i = 0; i < liRatings.Count; ++i)
            {
                liFzMatrix[0].Add((float)fzRating[liRatings[i]]);
            }

            for (int i = 1; i < fzRating.Count; ++i)
            {
                liFzMatrix.Add(new List<float>());
                for (int j = 0; j<fzRating.Count; ++j)
                {
                    float fTempRatingR = (float)fzRating[liRatings[i]];
                    float fTempRatingC = (float)fzRating[liRatings[j]];
                    liFzMatrix[i].Add(fTempRatingR/fTempRatingC);
                }
            }
            return liFzMatrix;
        }

        protected float calculateColumnSum(int iColNum)
        {
            float fSum = new float();
            for(int i = 0; i<liFzMatrix[iColNum].Count; ++i)
            {
                fSum += liFzMatrix[iColNum][i];
            }
            return fSum;
        }
        public ArrayList calculateBelongValues()
        {
            for(int i = 0; i< fzRating.Count; ++i)
            {
                float fColSum = this.calculateColumnSum(i);
                vBelongValues.Add((1 / fColSum) / liFzMatrix[liFzMatrix.Count-1].Max());
            }  
            return vBelongValues;
        }
        public void displayGraph()
        {

        }
             
    }

    public enum eAdvantages
    {
        APSEND = 1,
        INTERMEDIATE_WEAK = 2,
        WEAK = 3,
        INTERMEDIATE_ESSENTIAL = 4,
        ESSENTIAL = 5,
        INTERMEDIATE_APPARENT = 6,
        APPARENT = 7,
        INTERMEDIATE_ABSOLUTE = 8,
        ABSOLUTE = 9
    }

    public enum eRating 
    {
        LOW = 50,
        LIGHT_MIDDLE = 60,
        MIDDLE = 70,
        LIGHT_HIGH = 80,
        HIGH = 90,
        MAXIMAL = 100
    }

    public class TempRow
    {
        public float _50 { get; set; }
        public float _60 { get; set; }
        public float _70 { get; set; }
        public float _80 { get; set; }
        public float _90 { get; set; }
        public float _100 { get; set; }

    }
}

File: ai/labs/lab2/fuzzy_set/ratingfuzzyset.cpp
#include "ratingfuzzyset.h"

RatingFuzzySet::RatingFuzzySet()
{ 
    mapValues.insert(std::make_pair(LOW, APSEND));
    mapValues.insert(std::make_pair(LIGHT_MIDDLE, WEAK));
    mapValues.insert(std::make_pair(MIDDLE, ESSENTIAL));
    mapValues.insert(std::make_pair(LIGHT_HIGH, APPARENT));
    mapValues.insert(std::make_pair(HIGH, INTERMEDIATE_ABSOLUTE));
    mapValues.insert(std::make_pair(MAXIMAL, ABSOLUTE));
}

map<eRating, eAdvantages> RatingFuzzySet::getMapValues() const
{
    return this->mapValues;
}

array<array<float, ELEMENTS>, ELEMENTS> RatingFuzzySet::calculateMtxVls()
{
    array<array<float, ELEMENTS>, ELEMENTS> arMtx;
    array<int, ELEMENTS> arRatings(toArray());
    for(int i = 0; i<ELEMENTS; ++i){
        for(int j = 0; j<ELEMENTS; ++j){
            int iAssesR = mapValues.at((eRating)arRatings[i]);
            int iAssesC = mapValues.at((eRating)arRatings[j]);
            arMtx[i][j] = (float)iAssesR/(float)iAssesC;
        }
    }
    this->arMatrixFS = std::move(arMtx);
    return this->arMatrixFS;
}

array<int, ELEMENTS>&& RatingFuzzySet::toArray() const
{
    array<int, ELEMENTS> arrRating;
    arrRating[0] = LOW;
    arrRating[1] = LIGHT_MIDDLE;
    arrRating[2] = MIDDLE;
    arrRating[3] = LIGHT_HIGH;
    arrRating[4] = HIGH;
    arrRating[5] = MAXIMAL;
    return std::move(arrRating);
}

array<float, ELEMENTS>&& RatingFuzzySet::calculateColSum()
{
    array<float, ELEMENTS> arrSums;
    for(int i = 0; i<ELEMENTS; ++i){
        for(int j = 0; j<ELEMENTS; ++j){
            arrSums[i] += this->arMatrixFS[j][i];
        }
    }
    return std::move(arrSums);
}

map<eRating, float> RatingFuzzySet::calculateBlngValues()
{
    map<eRating, float> mapBlngVals;
    array<float, ELEMENTS> arrSums(calculateColSum());
    map<eRating, eAdvantages>::iterator itMap = this->mapValues.begin();
    for(int i = 0; i<ELEMENTS; ++i){
        mapBlngVals.insert(std::make_pair(itMap->first, (1.0/arrSums[i]) ));
        ++itMap;
    }
    return mapBlngVals;
}
